From c57970c07868adbfe4a1f2b1cb11709e9d85e118 Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Thu, 9 Apr 2020 14:35:26 +0800
Subject: [PATCH] usb: gadget: uvc: configfs: fix frame index and H264 format

This patch fixes the issue that "bFrameindex" attribute was
always set to "1". And add V4L2_PIX_FMT_H264 in uvc_formats[]
to support H264 format.

Fixes: 66cb0608ac27 ("FROMLIST: usb: gadget: uvc: configfs: add framebased format support")
Change-Id: Icd7ce96b9667c378de421fd714a42f243b543764
Signed-off-by: William Wu <william.wu@rock-chips.com>
---
 drivers/usb/gadget/function/uvc_configfs.c | 2 +-
 drivers/usb/gadget/function/uvc_v4l2.c     | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/gadget/function/uvc_configfs.c b/drivers/usb/gadget/function/uvc_configfs.c
index ec03b4933682..1a529b0f7a9b 100644
--- a/drivers/usb/gadget/function/uvc_configfs.c
+++ b/drivers/usb/gadget/function/uvc_configfs.c
@@ -1448,7 +1448,7 @@ static void uvcg_format_set_indices(struct config_group *fmt)
 	list_for_each_entry(ci, &fmt->cg_children, ci_entry) {
 		struct uvcg_frame *frm;
 
-		if (ci->ci_type != &uvcg_frame_type1 ||
+		if (ci->ci_type != &uvcg_frame_type1 &&
 		    ci->ci_type != &uvcg_frame_type2)
 			continue;
 
diff --git a/drivers/usb/gadget/function/uvc_v4l2.c b/drivers/usb/gadget/function/uvc_v4l2.c
index 097ee1adee8d..b25618b2ee04 100644
--- a/drivers/usb/gadget/function/uvc_v4l2.c
+++ b/drivers/usb/gadget/function/uvc_v4l2.c
@@ -58,6 +58,7 @@ struct uvc_format {
 static struct uvc_format uvc_formats[] = {
 	{ 16, V4L2_PIX_FMT_YUYV  },
 	{ 0,  V4L2_PIX_FMT_MJPEG },
+	{ 0,  V4L2_PIX_FMT_H264  },
 };
 
 static int
-- 
2.35.3

